addons/hr_timesheet/models/hr_timesheet.py

```diff
@@ -328,7 +328,14 @@ def _timesheet_preprocess(self, vals_list):
             if not vals.get('partner_id'):
                 vals['partner_id'] = data.partner_id.id
             if not vals.get('product_uom_id'):
-                vals['product_uom_id'] = uom_id_per_company.get(account_per_id[vals['account_id']].company_id, False)
+                company_id = account_per_id[vals['account_id']].company_id
+                if not company_id:
+                    task = self.env['project.task'].browse(vals.get('task_id'))
+                    company_id = task.analytic_account_id.company_id or task.company_id
+                if not company_id:
+                    project = self.env['project.project'].browse(vals.get('project_id'))
+                    company_id = project.analytic_account_id.company_id or project.company_id
+                vals['product_uom_id'] = uom_id_per_company.get(company_id, False) or self.env['res.company'].browse(company_id.id).project_time_mode_id.id
```

Comment:
As @auon-odoo said me, we could use `data` variable to search the company to use to avoid calling browse. 🙂 
And so we should have something like this:
```suggestion
                company = account_per_id[vals['account_id']].company_id
                if not company:
                    company = data.company_id
                vals['product_uom_id'] = uom_id_per_company.get(company.id, company.project_time_mode_id.id)
```
Is it right like that? @auon-odoo 🤔

Reply:
Thanks for the great feedback, I've implemented your suggestion, it's way less verbose, I don't know how I missed `data`.